lib/core/src/Cardano/Wallet/DB/Checkpoints.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE DeriveGeneric #-}
+
+-- |
+-- Copyright: © 2021 IOHK
+-- License: Apache-2.0
+--
+-- Data type that represents a collection of checkpoints.
+-- Each checkpoints is associated with a 'SlotPoint'.
+
+module Cardano.Wallet.DB.Checkpoints
+    ( getPoint
+
+    -- * Checkpoints  
+    , Checkpoints (..)
+    , singleton
+    , getLatest
+    , findNearestPoint
+    
+    -- * Delta types
+    , DeltaCheckpoints (..)
+    , DeltaMap (..)
+    ) where
+
+import Prelude
+
+import Data.Delta
+    ( Delta (..) )
+import Data.Generics.Internal.VL.Lens
+    ( over, view )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Map.Strict
+    ( Map )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Wallet.Primitive.Model as W
+import qualified Cardano.Wallet.Primitive.Types as W
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+
+{- NOTE [PointSlotNo]
+
+'SlotNo' cannot represent the genesis point.
+
+Historical hack. The DB layer can't represent 'Origin' in the database,
+instead we have mapped it to 'SlotNo 0', which is wrong.
+
+Rolling back to SlotNo 0 instead of Origin is fine for followers starting
+from genesis (which should be the majority of cases). Other, non-trivial
+rollbacks to genesis cannot occur on mainnet (genesis is years within
+stable part, and there were no rollbacks in byron).
+
+Could possibly be problematic in the beginning of a testnet without a
+byron era. /Perhaps/ this is what is happening in the
+>>> [cardano-wallet.pools-engine:Error:1293] [2020-11-24 10:02:04.00 UTC]
+>>> Couldn't store production for given block before it conflicts with
+>>> another block. Conflicting block header is:
+>>> 5bde7e7b<-[f1b35b98-4290#2008]
+errors observed in the integration tests.
+
+The issue has been partially fixed in that 'rollbackTo' now takes
+a 'SlotPoint' as argument, which can represent the 'Origin'.
+However, the database itself mostly stores slot numbers.
+
+FIXME LATER during ADP-1043: As we move towards in-memory data,
+all slot numbers in the DB file will either be replaced by
+the 'SlotPoint' type, or handled slightly differently when it
+is clear that the data cannot exist at the genesis point
+(e.g. for TxHistory).
+
+-}
+
+-- | Helper function: Get the 'Point' of a wallet state.
+getPoint :: W.Wallet s -> W.SlotPoint
+getPoint =
+    W.toSlotPoint . W.chainPointFromBlockHeader . view #currentTip
```

Comment:
what is returned when not even genesis block header? :thinking:

Reply:
I don't understand, what do you mean? If the current tip is not the genesis block header, then `At (Slot slot)` will be returned, where `slot` is the purported slot number recorded in the `BlockHeader`.

Reply:
yes, yes